google/cloud/pubsublite/internal/stream_factory.h

```diff
@@ -48,14 +49,16 @@ inline std::unique_ptr<grpc::ClientContext> MakeGrpcClientContext(
 inline StreamFactory<pubsublite::v1::PublishRequest,
                      pubsublite::v1::PublishResponse>
 MakeStreamFactory(
-    // NOLINTNEXTLINE(performance-unnecessary-value-param)
-    std::shared_ptr<pubsublite::v1::PublisherService::StubInterface> stub,
+    std::shared_ptr<pubsublite::v1::PublisherService::StubInterface> const&
```

Comment:
Maybe we should change things a bit.  Consider defining:

```cc
using StreamFactory =
    std::function<std::unique_ptr<BidiStream<Request, Response>>(
       google::cloud::CompletionQueue const& cq, std::unique_ptr<grpc::ClientContext> context)>;
```

Then you can define:

```cc
inline StreamFactory<pubsublite::v1::PublishRequest, pubsublite::v1::PublishResponse>
MakeStreamFactory(std::shared_ptr<pubsublite::v1::PublisherService::StubInterface> s) {
  return [stub = std::move(s)](
      google::cloud::CompletionQueue const& cq, std::unique_ptr<grpc::ClientContext> context) {
         return MakeStreamingReadWriteRpc(cq, std::move(context)
             [stub](grpc::ClientContext* context, grpc::CompletionQueue* cq) {
                 return stub->PrepareAsyncSubscribe(context, cq);
             }
      };
}
```

And you can define a generic decorator:

```cc
template <typename Request, typename Response>
inline StreamFactory<Request, Response>
AuthDecorator(
  std::shared_ptr<google::cloud::internal::GrpcAuthenticationStrategy> a,
  StreamFactory<Request, Response> c) {
  return [auth = std::move(s), child = std::move(c)](
    google::cloud::CompletionQueue const& cq, std::unique_ptr<grpc::ClientContext> context) {
      return absl::make_unique<AsyncStreamingReadWriteRpcAuth<Request, Response>>(
          cq, std::move(context), std::move(child));
    };
}
```

Reply:
You can implement the move-captures in C++11 using a helper struct inside the function, or just make more copies.

Reply:
I've restructured this to not duplicate lambda definitions. I still want the interface to have all the arguments pre-bound, but there's less code duplication now and less tight coupling.